Update tilemap interpretor.

calebkrauter opened the pull request:
Added Surface Snow
Moved tile map interpreter to its own class.
Add hill zone.
Add woods zone.
Add dark clouds.

Commit:
Refactor to weather system and precipitation. Add new zone.

Commit:
Add start for outside cave.

Add surface snow and starter tilemap.

Commit:
Implement easter egg.

Commit:
Add hill down from village.

Add more blocks and a new zone.

Commit:
Implement new zones.

Add hill zone.
Add woods zone.
Remove some unnecessary assets.
Add some new sprites. (Add dark clouds.)

Commit:
Update to Chad's sprite sheet.

Fix issue where chad's animation was clipping.

calebkrauter commented on js/scripts/util.js:
I added a little easteregg here.

coltenkrauter commented on js/scripts/util.js:
```diff
@@ -271,7 +281,48 @@ const EVENT_HANDLERS = {
+        }
+        console.log(mickeyCount);
+        if (mickeyCount == 6) {
+            window.location.href = "https://www.youtube.com/watch?v=hmzO--ox7X0";
```
This is fun. Will it cause the game progress to be lost.. unsaved?

tclaar replied:
Oh, good point. @calebkrauter , maybe open this in a new window/tab?

calebkrauter replied:
Thanks for pointing this out! We can either do a save before the link used or open in a new tab which is likely possible. @B1GD3VD0G how difficult would it be to just add another save at this point?

calebkrauter replied:
For now I made it open a new page.

coltenkrauter commented on js/scripts/zone_loaders/tilemap_interpreter.js:
```diff
@@ -0,0 +1,153 @@
+                    case 'c':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.PLANKS_REDWOOD_LIGHT));
+                        break;
+                    case 'b':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.PLANKS_REDWOOD));
+                        break;
+                    case 'a':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.LOG_SPRUCE_VIRTICAL));
+                        break;
+                    case '9':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.LOG_SPRUCE_HORIZONTAL));
+                        break;
+                    case '8':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.PLANKS_SPRUCE));
+                        break;
+                    case '7':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.PLANKS_OAK));
+                        break;
+                    case '6':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.BARS));
+                        break;
+                    case '5':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.LAVA_ROCK));
+                        break;
+                    case '4':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.SNOWY_ICE));
+                        break;
+                    case '3':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.SNOWY_DIRT));
+                        break;
+                    case '2':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.GRASS));
+                        break;
+                    case '1':
+                        GAME.addEntity(new Block(new Vector(x, y), Block.DIRT), -1);
+                        break;
+
+                    default:
+                        break;
+                }
```
This could be simplified. Generally, if a switch is made up of one-line blocks, I prefer a different approach like building a map as you have done elsewhere.

calebkrauter replied:
Got it, thanks! I was wondering about that and am not a fan of this approach either.

calebkrauter replied:
I was running into issues trying to make this change. For some reason this is only making the dirt blocks. It should be interpreting all the blocks and putting them in the midground and after I figure that out i'd sort out which go to what ground and put surface_snow down only if it is snowing. I found that if I set the conditional near the bottom to != instead of == it displayed the inverse meaning all the blocks got displayed save for the dirt blocks and it seemed like air blocks with bounding boxes were made as well. Can you see what I'm missing? For now I'll finish pushing my other changes and come back to this if time permits.
<img width="348" alt="Screenshot 2024-03-01 at 12 51 08 AM" src="https://github.com/GoodBadChad/good-bad-chad-br/assets/42187569/26b74107-2998-4824-9df5-734c6adbd4a8">
<img width="547" alt="Screenshot 2024-03-01 at 12 49 23 AM" src="https://github.com/GoodBadChad/good-bad-chad-br/assets/42187569/b74066d8-d268-4de3-9cfd-ded8193b7fa2">

tclaar commented on js/classes/entities/midground/block.js:
```diff
@@ -97,6 +97,51 @@ class Block {
+    }
+    static get HIDDEN_BLOCK_CLOUD() {
+        return 35;
+    }
```
What do all of these constants mean? JSDoc would be helpful.

calebkrauter replied:
My bad! I'll update that. Hidden block is for putting in places that don't normally have blocks. Like hiding a bounding box to step on behind a cloud for example.

tclaar commented on js/scripts/zone_loaders/tilemap_interpreter.js:
Thank you for providing this!

calebkrauter replied:
No problem!

tclaar reviewed (approved):
Looks good! Thanks Caleb.

Commit:
Implement Cave zone.

Update to sprites.

Commit:
Update to cave

Fixes a place where chad couldn't get go back up to the start.

Commit:
Fixed an issue where Chad got stuck.

Commit:
Add in lanterns.

Bug having issues adding in rain or snow to main village but works elsewhere.

tclaar commented on js/classes/entities/foreground/weather_system.js:
```diff
@@ -74,35 +79,37 @@ class WeatherSystem {
 
-
+                let cloudTypeOffset = 0;
+                if (weatherType.RAIN) {
```
This is always going to evaluate to true, since `weatherType.RAIN` always exists.

calebkrauter replied:
Oh, interesting. Thanks.

tclaar commented on js/classes/zone.js:
```diff
@@ -135,33 +135,22 @@ class Zone {
-            },
-            mountain: {
-
-            },
```
I don't think it's a good idea to put all of the zones under village and delete these categories. Having categories like `village`, `mountain`, `forest`, etc will keep it much more organized.

calebkrauter replied:
We can add those back, I was thinking that it would make sense to get rid of them because we won't have that many zones.

tclaar commented on js/scripts/zone_loaders/village_zone_loaders.js:
```diff
@@ -415,22 +304,30 @@ const loadVillageMain = () => {
+            surfaceSnow = true
+        }
+        new tilemapInterpreter(villageMainTileMap, surfaceSnow);
+        new WeatherSystem(weather, 2, "day");
```
Why are these constructors? They are being evoked essentially as methods and the constructed instances are never used. Unless you expect to use either of these as objects, these should just be static methods.

calebkrauter replied:
Gotcha, why is that? I can change that, I don't understand the benefit of changing it. That being said I also don't have an issue with changing it.

tclaar replied:
The main reason is that you're not using the resulting instance, which is the whole point of a constructor. Calling a constructor without using the result is essentially the same thing as using a static method.

calebkrauter replied:
I am concerned that changing it to using static methods and referencing the class in a static context will cause issues. If you set weather to rain in one zone and the zone is disposed of and the new zone sets the weather to snow then would it not cause rain and snow to both fall because of the static reference? Inside of the weather system entities are added and if we make the function call unnecessarily multiple times add the weather after setting it we would be adding in the rain or snow if I'm not mistaken and it will never remove the rain or snow when it needs to as the function calls are not static references. Is there something missing in my logic? I'll take a look and see if I can make the change you're suggesting however I'll be working all weekend and would suggest that you commit the changes you are suggesting to this PR if I am unable to complete the changes or convince you otherwise in a timely manner.

calebkrauter replied:
I made the change. You were correct that this works however I don't understand what issue there is with instances in this situation. I'll push my changes soon. Thanks for the feedback!

tclaar replied:
Thanks for making the change. It's certainly not a big deal, and like I said it wasn't causing any issues functionality wise as it was, but if you were using a constructor as if it was a static method (which you were), why not use a static method? The *entire* point of a constructor is to create a new instance. If you don't need an instance, you shouldn't be using a constructor. To me, it's bad practice in general to use a programming construct improperly like that without a good reason, because it can be unclear to others reading your code what you're trying to do, lead to poor design choices, and potentially have unintended side effects caused by the functionality of the specific construct.

tclaar reviewed (changes requested):
I've been working with your code (and thank you! it is helpful) and have some new thoughts.

calebkrauter committed:
Update weather_system.js

Updated based on feedback.

Commit:
Use static ref instead of instances based on @tclaar and @NathanHinthorne 's feedback

Commit:
Add back zone loader classes per @tclaar 's request.

I doubt these are really necessary with our current scope and we can decide later if we would like to remove them. As you have suggested I have added these back from a previous PR, they are merely empty files as I found them in another PR.

Commit:
Update doc and easter egg based on feedback.

calebkrauter commented:
I have updated my PR based on all feedback save for updating the switch statement to a map. While I would like to make this change at some point, due to issues and time I propose we merge this PR as is. The switch statement as is does not pose threats to current code functionality though it is messy and all other points of feedback have been addressed.

tclaar commented on js/classes/entities/foreground/weather_system.js:
```diff
@@ -1,52 +1,49 @@
+        WeatherSystem.isRain = WeatherSystem.type === weatherType.RAIN;
+        WeatherSystem.isSnow = WeatherSystem.type === weatherType.SNOW;
+        WeatherSystem.isClouds = WeatherSystem.type === weatherType.CLOUDS;
+        WeatherSystem.startingYOfClouds = startingYOfClouds;
```
You shouldn't be using static fields - you don't need anything to persist. Instead, just pass these values as parameters to the different methods that need them.

tclaar commented on js/classes/entities/midground/block.js:
Thanks for the JSDoc!

NathanHinthorne commented on js/classes/entities/background/decoration.js:
Love these new trees!

Commit:
Update to lanterns.

Commit:
Update to lanterns in cave.

NathanHinthorne reviewed (approved):
Great changes to the weather and tilemap reader. Thanks @calebkrauter !
